### ISSUE TITLE
gdnametest (HEAD) fails test on OS X (case sensitivity?)

### ISSUE THREAD

**Issue by nieder** [libgd/libgd#385]
On OS X 10.11.6, gdimagefile/gdnametest fails. This is the main build log:
~~~~
PASS: gdimagecrop/bug00297
PASS: gdimagecrop/php_bug_72494
FAIL: gdimagefile/gdnametest
PASS: gdimagefill/bug00002_1
PASS: gdimagefill/bug00002_2

~~~~

From tests/test-suite.log:
~~~~
========================================
   GD 2.3.0-dev: tests/test-suite.log
========================================

# TOTAL: 161
# PASS:  160
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: gdimagefile/gdnametest
============================

gdimagefile/gdnametest.c:122: 255 pixels different on /var/folders/h1/_gcxvqbd23b52kxbnrrh3qmc0000gn/T//gdtest.nlivFF/img.gif
gdimagefile/gdnametest.c:122: 255 pixels different on /var/folders/h1/_gcxvqbd23b52kxbnrrh3qmc0000gn/T//gdtest.nlivFF/img.GIF
FAIL gdimagefile/gdnametest (exit status: 2)
~~~~

All tests passed with 2.2.3 and fail with 2.2.4 and up. The machine is running in a case-insensitive filesystem (the OS X default), so I wonder if `img.gif` and `img.GIF` are causing issues there.

**Comment by cmb69** [comment]
> The machine is running in a case-insensitive filesystem (the OS X default), so I wonder if img.gif and img.GIF are causing issues there.

I don't think so, but it would be easy to try out: just delete [this line](https://github.com/libgd/libgd/blob/gd-2.2.4/tests/gdimagefile/gdnametest.c#L64). I don't have a macOS machine available, but at least on Windows NTFS (case-insensitive as well) the test succeeds for me.

I guess the issue is already stated in the comment in the [line above](https://github.com/libgd/libgd/blob/gd-2.2.4/tests/gdimagefile/gdnametest.c#L63): "This seems to come from tc<->palette". While it is not clear, what is meant by "this", I suppose that it's about the margin-of-error ("Maximum total pixel diff"). The read-write tests create truecolor images, but saving as GIF converts to palette images (involving system dependend color quantization) and even dithering. This is likely to result in differences, and the margin-of-error (5 in this case) appears to be much too small for your system.

Actually, I wonder why this test [uses mkcross()](https://github.com/libgd/libgd/blob/gd-2.2.4/tests/gdimagefile/gdnametest.c#L104) – using mkwhite() instead should be perfectly fine for this test (which is about reading/writing the appropriate formats). And perhaps the test would be even more resilient if a palette image would be created in mkwhite().

**Comment by nieder** [comment]
Removing line 64 still causes the 'img.gif' test to fail. So not case-sensitivity: both img.gif and img.GIF fail.

Not surprising from the error log, the culprit is [here](https://github.com/libgd/libgd/blob/gd-2.2.4/tests/gdimagefile/gdnametest.c#L123). Commenting that out lets the test pass.

Happy to test any patches.

**Comment by cmb69** [comment]
Thanks. I wonder how `img.gif` looks on your system, so please test with the following patch:
````
 tests/gdimagefile/gdnametest.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/tests/gdimagefile/gdnametest.c b/tests/gdimagefile/gdnametest.c
index b6b7c56..099878d 100644
--- a/tests/gdimagefile/gdnametest.c
+++ b/tests/gdimagefile/gdnametest.c
@@ -105,7 +105,7 @@ do_test() {
         if (!names[n].readonly) {
             /* Prepend the test directory; this is expected to be run in
              * the parent dir. */
-            full_filename = gdTestTempFile(names[n].nm);
+            full_filename = gdTestFilePath2("gdimagefile", names[n].nm);
             status = gdImageFile(orig, full_filename);
             gdTestAssertMsg(status == GD_TRUE, "Failed to create %s\n", full_filename);
         } else {
@@ -122,8 +122,8 @@ do_test() {
         gdTestAssertMsg(pixels <= names[n].maxdiff, "%u pixels different on %s\n", pixels, full_filename);
 
         if (!names[n].readonly) {
-            status = remove(full_filename);
-            gdTestAssertMsg(status == 0, "Failed to delete %s\n", full_filename);
+            /* status = remove(full_filename);
+               gdTestAssertMsg(status == 0, "Failed to delete %s\n", full_filename); */
         }/* if */
 
         free(full_filename);
````
This is supposed to create `tests/gdimagefile/img.gif`. Please post the file here.

**Comment by nieder** [comment]
![img](https://user-images.githubusercontent.com/1105137/29756514-3068c072-8b6a-11e7-861e-681ca35b09c3.gif)
Notice the two stray white pixels on left bar.

**Comment by cmb69** [comment]
Thanks, that was quite helpful. For reference, this is the image that I get:
![img](https://user-images.githubusercontent.com/2306138/29769609-730352a4-8beb-11e7-8e86-407fd674ed7c.gif)
Indeed, only two pixels differ, and having a closer look at gdMaxPixelDiff() shows that this function doesn't return the number of different pixels, but rather the largest pixel difference. Obviously, the error message is plain wrong.

I still wonder, though, why you're getting these two stray white pixels. I assume this is due to dithering, but I get the same whether I've build GD with or without libimagequant (`--with-liq`). Maybe there is a particular issue with pixels at the border of the image (besides that dithering appears to be broken anyway, see #293), so please modify [mkcross()](https://github.com/libgd/libgd/blob/gd-2.2.4/tests/gdimagefile/gdnametest.c#L39-L40) to not to draw to the very edges:
````
        gdImageLine(im, LX-n, 1, LX-n, HEIGHT-2, fg);
        gdImageLine(im, 1, LY-n, WIDTH-2, LY-n, fg);
````
This gives me:
````
gdimagefile/gdnametest.c:122: 255 pixels different on ./gdimagefile/img-ref.xpm
gdimagefile/gdnametest.c:122: 255 pixels different on ./gdimagefile/img-ref.xbm
gdimagefile/gdnametest.c:122: 255 pixels different on ./gdimagefile/img-ref.tga
```` 
what is to be expected, since these images are compared to prepared files.

**Comment by nieder** [comment]
~~~~
gdimagefile/gdnametest.c:125: 255 pixels different on ./gdimagefile/img-ref.xpm
Skipping test for 'img-ref.xbm'.  FIX THIS!
Skipping test for 'img-ref.tga'.  FIX THIS!
Skipping test for 'img.webp'.  FIX THIS!
Skipping test for 'img.tiff'.  FIX THIS!
FAIL gdimagefile/gdnametest (exit status: 1)
~~~~
The created img.gif looks like this:
![img](https://user-images.githubusercontent.com/1105137/29903845-ae6af1da-8dcb-11e7-8c76-fafed1be2ccc.gif)
No stray pixels, and the edge part of each arm of the cross is not drawn.

**Comment by cmb69** [comment]
> No stray pixels, and the edge part of each arm of the cross is not drawn.

Fine, thanks! I shall modify gdnametest respectively, and add a separate test for the broken GIF behavior.

**Comment by cmb69** [comment]
@nieder Most recently I've stumbled upon #409, which **might** be the reason that the test is failing for you. Can you please verify with `CFLAGS=-DRIGHT_SHIFT_IS_UNSIGNED ./configure …`?

**Comment by nieder** [comment]
With `-DRIGHT_SHIFT_IS_UNSIGNED` in CFLAGS:
~~~~
gdimagefile/gdnametest.c:125: 255 pixels different on /tmp/gdtest.5Q5ejO/img.gif
gdimagefile/gdnametest.c:125: 255 pixels different on /tmp/gdtest.5Q5ejO/img.GIF
~~~~
![img](https://user-images.githubusercontent.com/1105137/30002246-0b188c30-906a-11e7-848a-6e360c33902a.gif)

**Comment by cmb69** [comment]
Thanks, so `-DRIGHT_SHIFT_IS_UNSIGNED` doesn't make a difference here. I still wonder where these two white pixels come from.

**Comment by pierrejoye** [comment]
@nieder good morning :)

Old bug, however I like to try again.

None of us has mac to try. We setup actions with MacOS but I totally fail to get the tests running (conflicts between cmake detection using brew deps and xcode libs). I am happy to add whatever is needed to get it compiled and run smoothly but I will need some help here.

Do you have a flow to get that build&test run smoothly? Our config for the MacOs action can be read here:
.github\workflows\ci_master.yml

The ticket about tests on mac is https://github.com/libgd/libgd/issues/732

**Comment by pierrejoye** [comment]
Sidenote, I am not sure yet why, but that exact test does not run on MacOS yet, while WEBP, TIFF and PNG are enabled, as well as the GD/GD2 format.

Debugging and let see.

**Comment by pierrejoye** [comment]
Start  62: test_gdimagefile_gdnametest
 62/211 Test  #62: test_gdimagefile_gdnametest ...................................   Passed    0.13 sec

MacOs 10.15, Xcode 12.3

See the github actions MacOS environment. So I suppose it is about case sensitivity after all.

Please reopen if you see alternative to make it work on your system. However as of now, MacOS tests pass.

**Comment by nieder** [comment]
Sorry, lost track of this.
Unfortunately, gdimagefile/gdnametest now passes for me as well. My system is now macOS10.14 and gd-2.3.2. It's the same physical hardware after an OS upgrade (still case insensitive) and upgrade in all the supporting libraries (libpng, libtiff, imagequant, etc).
